druid/src/main/java/org/apache/calcite/adapter/druid/DruidQuery.java

```diff
@@ -1124,24 +1130,49 @@ public String getQueryString(String pagingIdentifier, int offset) {
         }
       }
       this.timeZone = timeZone;
+      this.dateFormatter = new SimpleDateFormat(TimeExtractionFunction.ISO_TIME_FORMAT,
+          Locale.ROOT);
+      if (timeZone != null) {
+        this.dateFormatter.setTimeZone(TimeZone.getTimeZone(timeZone));
+      }
     }
 
     protected void clearFieldNameLists() {
       dimensions.clear();
       metrics.clear();
     }
 
-    @SuppressWarnings("incomplete-switch") String translate(RexNode e, boolean set) {
+    @SuppressWarnings("incomplete-switch")
+    /**
+     * formatDateString is used to format timestamp values to druid format using
+     * {@link DruidQuery.Translator#dateFormatter}. This is needed when pushing timestamp
+     * comparisons to druid using TimeFormatExtractionFunction that returns a string value.
+     */
+    String translate(RexNode e, boolean set, boolean formatDateString) {
       int index = -1;
       switch (e.getKind()) {
       case INPUT_REF:
         final RexInputRef ref = (RexInputRef) e;
         index = ref.getIndex();
         break;
       case CAST:
-        return tr(e, 0, set);
+        return tr(e, 0, set, formatDateString);
```

Comment:
To avoid incorrect results, we should change:
https://github.com/nishantmonu51/calcite/blob/82ae538f861e485ee4b1a47762353f58e83c3047/druid/src/main/java/org/apache/calcite/adapter/druid/DruidQuery.java#L271
to make it a bit more restrictive. In particular, we can make it similar to the check for casts within ```literalValue``` method but without the nullability check. What do you think?

Reply:
done.